File: src/ships/purchaseScanner.ts
import type { ScannerCandidate } from "../data/types/scannerCandidate.ts";
import type { ScannerPool } from "../data/types/scannerPool.ts";
import type { Scanner } from "../data/types/scanner.ts";
import type { Wallet } from "../data/types/wallet.ts";
import type { PurchaseScannerResult } from "../data/types/purchaseScannerResult.ts";
import { SCANNER_PURCHASE_COST_BY_TIER } from "../data/constants/shipsAndTravelConfig.ts";

// Scanner/Probe GDD §2.2. Mirrors purchaseShip() exactly: takes the actual
// candidate/pool/wallet data directly rather than IDs into an implicit
// store (same pure-function reasoning applied throughout this codebase),
// checks wallet sufficiency before deducting, removes the candidate from
// its pool, and transfers ownership.
export function purchaseScanner(
  candidate: ScannerCandidate,
  pool: ScannerPool,
  wallet: Wallet,
  playerId: string,
): PurchaseScannerResult {
  if (!pool.availableScanners.some((entry) => entry.id === candidate.id)) {
    return { purchased: false, reason: "candidate is not in this planet's scanner pool" };
  }

  const cost = SCANNER_PURCHASE_COST_BY_TIER.find((entry) => entry.tier === candidate.tier)?.cost;
  if (cost === undefined) {
    throw new RangeError(`no purchase cost defined for tier ${candidate.tier}`);
  }
  if (wallet.credits < cost) {
    return { purchased: false, reason: `insufficient funds: need ${cost}, have ${wallet.credits}` };
  }

  const scanner: Scanner = { id: candidate.id, tier: candidate.tier, ownerId: playerId };

  return {
    purchased: true,
    scanner,
    updatedPool: {
      ...pool,
      availableScanners: pool.availableScanners.filter((entry) => entry.id !== candidate.id),
    },
    updatedWallet: { ...wallet, credits: wallet.credits - cost },
  };
}

File: unity/ProfitableCore/Simulation/PurchaseScanner.cs
using Profitable.Core.Constants;
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/ships/purchaseScanner.ts. Mirrors PurchaseShip exactly.
public static class ScannerPurchaser
{
    public static PurchaseScannerResult PurchaseScanner(ScannerCandidate candidate, ScannerPool pool, Wallet wallet, string playerId)
    {
        if (!pool.AvailableScanners.Any(entry => entry.Id == candidate.Id))
        {
            return new PurchaseScannerRejected { Reason = "candidate is not in this planet's scanner pool" };
        }

        if (!ShipsAndTravelConfig.ScannerPurchaseCostByTier.TryGetValue(candidate.Tier, out var cost))
        {
            throw new InvalidOperationException($"no purchase cost defined for tier {candidate.Tier}");
        }
        if (wallet.Credits < cost)
        {
            return new PurchaseScannerRejected { Reason = $"insufficient funds: need {cost}, have {wallet.Credits}" };
        }

        var scanner = new Scanner { Id = candidate.Id, Tier = candidate.Tier, OwnerId = playerId };

        var updatedPool = new ScannerPool
        {
            PlanetId = pool.PlanetId,
            AvailableScanners = pool.AvailableScanners.Where(entry => entry.Id != candidate.Id).ToList(),
            LastRefreshedAt = pool.LastRefreshedAt,
        };

        return new PurchaseScannerSucceeded
        {
            Scanner = scanner,
            UpdatedPool = updatedPool,
            UpdatedWallet = new Wallet { PlayerId = wallet.PlayerId, Credits = wallet.Credits - cost },
        };
    }
}
